rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -1628,25 +1655,235 @@ rmw_return_loaned_message_from_subscription(
+    [client_data]() {
+      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
+        client_data->response_type_support->~ResponseTypeSupport(),
+        ResponseTypeSupport);
+    });
+
+  // Populate the rmw_client.
+  rmw_client->data = client_data;
+  rmw_client->implementation_identifier = rmw_zenoh_identifier;
+
+  rmw_client->service_name = rcutils_strdup(service_name, *allocator);
+
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    rmw_client->service_name,
+    "failed to allocate client name",
+    return nullptr);
+
+  auto free_service_name = rcpputils::make_scope_exit(
+    [rmw_client, allocator]() {
+      allocator->deallocate(const_cast<char *>(rmw_client->service_name), allocator->state);
+    });
+
+  // Zenoh implementation for the client
+
+  // TODO(francocipollone): Replace ros_topic_name_to_zenoh_key by service related function.
+  //                        If this is enough simply rename the method.
+  z_owned_keyexpr_t keyexpr = ros_topic_name_to_zenoh_key(
+    rmw_client->service_name, node->context->actual_domain_id, allocator);
+  auto always_free_ros_keyexpr = rcpputils::make_scope_exit(
+    [&keyexpr]() {
+      z_keyexpr_drop(z_move(keyexpr));
+    });
+  if (!z_keyexpr_check(&keyexpr)) {
+    RMW_SET_ERROR_MSG("unable to create zenoh keyexpr.");
+    return nullptr;
+  }
+
+  free_rmw_client.cancel();
+  free_client_data.cancel();
+  free_request_type_support.cancel();
+  destruct_request_type_support.cancel();
+  free_response_type_support.cancel();
+  destruct_response_type_support.cancel();
+  free_service_name.cancel();
+
+  return rmw_client;
 }
 
 //==============================================================================
 /// Destroy and unregister a service client from its node.
 rmw_ret_t
 rmw_destroy_client(rmw_node_t * node, rmw_client_t * client)
 {
-  static_cast<void>(node);
-  static_cast<void>(client);
-  return RMW_RET_UNSUPPORTED;
+  RCUTILS_LOG_DEBUG_NAMED("rmw_zenoh_cpp", "[rmw_destroy_client] %s", client->service_name);
+
+  // ASSERTIONS ================================================================
+  RMW_CHECK_ARGUMENT_FOR_NULL(node, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(client, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    node,
+    node->implementation_identifier,
+    rmw_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    client,
+    client->implementation_identifier,
+    rmw_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
+  rcutils_allocator_t * allocator = &node->context->options.allocator;
+
+  auto client_data = static_cast<rmw_client_data_t *>(client->data);
+
+  // CLEANUP ===================================================================
+  allocator->deallocate(client_data->request_type_support, allocator->state);
+  allocator->deallocate(client_data->response_type_support, allocator->state);
+  allocator->deallocate(client->data, allocator->state);
```

Comment:
Make sure to call `z_drop()` on `z_owned_XX` objects in `client_data` before deallocating

Reply:
Done.